### PR TITLE
Move to `make`

### PR DESCRIPTION
**Pull request by sndrs**
This is a proposal to elevate `make` to become the default way to manage the assets from the command line, retiring `grunt` (though not behind the scenes). 

It should cover all the day-to-day things for which we currently use `grunt`, and improve upon some of them (this is useful, for example, in the instance `grunt install` doesn't work because you haven't `npm install`ed `grunt` already; it also means no API changes to run the site whatever we use in the future).
- replaces `grunt install` and `grunt compile`
  screen shot 2015-10-29 at 16 59 51
- adds a `make watch` command that fires up a browser-sync instance and live-reloads `*.js`/`*.scss` changes (to take advantage of #11009)
  screen shot 2015-10-29 at 17 10 09
- adds a `make reinstall` to nuke 3rd party dependencies and reinstall them in one go
- adds a `make shrinkwrap` to easily lock down NPM packages
  screen shot 2015-10-29 at 18 56 43
- displays an options menu using just `make` or `make help`
  screen shot 2015-10-29 at 19 05 08

thoughts?

### CODE REVIEW

File: Gruntfile.js [hunk 1]
```diff
@@ -57,9 +57,9 @@ module.exports = function (grunt) {
         grunt.log.subhead('Running Grunt in DEV mode');
     }
 
-    // Default task
+    // Default task - used by grunt-tc
```
Comment (OliverJAsh) [inline Gruntfile.js]: Could TeamCity use make instead of grunt too? Not sure if that makes sense

File: dev/bs-config.js [hunk 2]
```diff
@@ -27,6 +27,7 @@ module.exports = {
         "cwd": "../"
     },
     "files": [
+        "static/src/javascripts/*.js",
         "static/src/javascripts/**/*.js",
```
Comment (OliverJAsh) [inline dev/bs-config.js]: Won't this glob match everything caught by the glob you added?

File: Gruntfile.js (hunk 1 above)
Comment (sndrs) [inline Gruntfile.js]: yeah potentially. i wanted to limit this to (hopefully) non-breaking stuff. assume `make` is available there etc, but was't sure. happy to change if anyone (@rich-nguyen?) can confirm it should be ok?

File: dev/bs-config.js (hunk 2 above)
Comment (sndrs) [inline dev/bs-config.js]: you'd think so! but it wasn't picking up top level (`core.js` etc) file changes without it...

File: makefile [hunk 3]
```diff
@@ -1,27 +1,56 @@
...
-	grunt test --dev
+	@grunt test --dev
 
 validate:
-	grunt validate
+	@grunt validate
+
+shrinkwrap:
+	@npm shrinkwrap && node dev/clean-shrinkwrap.js
+	@node dev/message.js shrinkwrap
```
Comment (OliverJAsh) [inline makefile]: Am I reading this wrong? When we run `make shrinkwrap`, we shrinkwrap and then show a warning message?

File: dev/bs-config.js (hunk 2 above)
Comment (OliverJAsh) [inline dev/bs-config.js]: Oh, must be a bug!

File: makefile (hunk 3 above)
Comment (sndrs) [inline makefile]: should just be a confirmation message?

**Comment by OliverJAsh** [comment]
I especially like `make help` :smile: Nice!

**Comment by sndrs** [comment]
Made the `grunt compile` message a bit more useful:

screen shot 2015-10-30 at 11 26 23

File: git-hooks/message.js [hunk 4]
```diff
@@ -5,10 +5,10 @@ var megalog = require('megalog');
  */
 switch (process.argv[2]) {
   case 'update':
-    megalog.info('Dependencies have been updated.\n\nRun `grunt install`.');
+    megalog.info('Dependencies have been updated.\n\nRun `make install`.');
     break;
 
   case 'shrinkwrap':
-    megalog.error('This commit changes `package.json`, but doesn\'t include `npm-shrinkwrap.json`.\n\nIf you\'re updating npm packages, please run `npm shrinkwrap && node dev/clean-shrinkwrap.js` to write the new dependencies to `npm-shrinkwrap.json` and add that file before committing.');
+    megalog.error('You have changes to `package.json` but havn’t updated `npm-shrinkwrap.json`.\n\nPlease run `make shrinkwrap` and commit the changes to `npm-shrinkwrap.json`.');
```
Comment (gtrufitt) [inline git-hooks/message.js]: Minor spelling on haven't

File: makefile [hunk 5]
```diff
@@ -1,27 +1,56 @@
...
+uninstall: uninstall-application uninstall-dev
+	@echo 'All 3rd party dependencies have been uninstalled.'
 
 test:
-	grunt test --dev
+	@grunt test --dev
 
 validate:
-	grunt validate
+	@grunt validate
```
Comment (gtrufitt) [inline makefile]: Would we want to be able to validate css and js separately via make like running ``grunt validate:css``? Same with compile? Not sure if it's needed at the higher level (or whether the watcher on compile compiles specific files?)

**Comment by rich-nguyen** [comment]
I'm happy to use `make`, `grunt`, `gulp`, etc..., whatever the wider team chooses they are most comfortable with maintaining. Be sure to communicate the rationale, benefits and the workflow to everyone who needs to be informed, especially the browsersync stuff.

I agree with @OliverJAsh, team city should build things a similar way to how we do it locally. It may seem less risky to skip this from a Pull Request perspective, but it's more risky in the long run to support several different ways of building.

Also if `make` is the principal build system, can the `dev/package.json` be moved into the top level `package.json` under `devDependencies`, and move all of  `dev/makefile` into it's parent `makefile`?

:+1:

File: makefile (hunk 5 above)
Comment (sndrs) [inline makefile]: yeah good thinking. not sure why github hasn't picked this up, but change for this is here: https://github.com/guardian/frontend/commit/46cd127f950bc21fd1e120bb540fbe563f05553e

Comment (gtrufitt) [inline makefile]: Aces 
yiss

**Pull request merged**